builder.go

```diff
@@ -0,0 +1,92 @@
+package cid
+
+import (
+	mh "github.com/multiformats/go-multihash"
+)
+
+type Builder interface {
+	Sum(data []byte) (*Cid, error)
+	GetCodec() uint64
+	WithCodec(uint64) Builder
+}
+
+type V0Builder struct{}
+
+type V1Builder struct {
+	Codec   uint64
+	HashFun uint64
+	HashLen int // HashLen <= 0 means the default length
+}
+
+func PrefixToBuilder(p Prefix) Builder {
+	if p.Version == 0 {
+		return V0Builder{}
+	}
+	mhLen := p.MhLength
+	if p.MhType == mh.ID {
+		mhLen = 0
+	}
+	if mhLen < 0 {
+		mhLen = 0
+	}
+	return V1Builder{
+		Codec:   p.Codec,
+		HashFun: p.MhType,
+		HashLen: mhLen,
+	}
+}
+
+func (p Prefix) GetCodec() uint64 {
```

Comment:
If `Prefix` is going to implement `Builder`, `PrefixToBuilder(p).Sum(...)` should be functionally equivalent to `p.Sum(...)`. My solution would be to just put the logic in `PrefixToBuilder` into `Prefix.Sum` but the alternative is to just *call* `PrefixToBuilder` in `Sum` (actually, that may just be the better solution).